test/test_optional.cpp

```diff
@@ -0,0 +1,629 @@
+// Copyright 2024 Man Group Operations Limited
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "sparrow/optional.hpp"
+
+#include "doctest/doctest.h"
+
+namespace sparrow
+{
+    using optional_double = optional<double, bool>;
+    using optional_int = optional<int, bool>;
+
+    TEST_SUITE("optional value")
+    {
+        TEST_CASE("constructors")
+        {
+            {
+                optional_double d;
+                CHECK_FALSE(d.has_value());
+            }
+
+            {
+                optional_double d(std::nullopt);
+                CHECK_FALSE(d.has_value());
+            }
+
+            {
+                optional_double d(1.2);
+                CHECK(d.has_value());
+                CHECK_EQ(d.value(), 1.2);
+            }
+
+            {
+                int i = 3;
+                optional_double d(i);
+                CHECK(d.has_value());
+                CHECK_EQ(d.value(), 3.);
+            }
+
+            {
+                double val = 1.2;
+                bool b1 = true;
+
+                optional_double td1(val, b1);
+                optional_double td2(std::move(val), b1);
```

Comment:
The std::move is useless as the val is not movable. Not sure about what you want to test

Reply:
So here we want to the the constructor overload that takes an rvalue reference. `std::move` does not actually "move" anything, it just casts its argument to an rvalue reference. Whether we have a moving constructor called in the end or not is not what we are intereested in here.